trace-forward/src/Trace/Forward/Utils/TraceObject.hs

```diff
@@ -23,129 +32,96 @@ import qualified Trace.Forward.Protocol.TraceObject.Forwarder as Forwarder
-        Just <$!> flushTBQueue q
+writeToSinkSTM :: TBQueue lo -> lo -> STM [lo]
+writeToSinkSTM queue traceObject = do
+    ---------- STM transaction: start ----------
+    isFull <- isFullTBQueue queue
+    !flushedTraceObjects <- if isFull
+                            then flushTBQueue queue
+                            else pure []
+    writeTBQueue queue traceObject
+    pure flushedTraceObjects
+    ---------- STM transaction: end ----------
 
-  -- if the sink was used and it
-  maybeShrinkQueue q = do
-    whenM (readTVar wasUsed) $ do
-      qLen <- lengthTBQueue q
-      when (fromIntegral qLen == disconnectedSize) $ do
-        -- trace ("shrinkQueue connected " ++ show connectedSize) $
-        switchQueue connectedSize
-
-  switchQueue size =
-    newTBQueue (fromIntegral size) >>= modifyTVar' forwardQueue . const
+--------------------------------------------------------------------------------
 
 readFromSink
   :: ForwardSink lo -- ^ The sink contains the queue we read 'TraceObject's from.
   -> Forwarder.TraceObjectForwarder lo IO ()
-readFromSink ForwardSink{forwardQueue, wasUsed} =
+readFromSink ForwardSink{forwardQueue} =
   Forwarder.TraceObjectForwarder
-    { Forwarder.recvMsgTraceObjectsRequest = \blocking (NumberOfTraceObjects n) ->
-        case blocking of
-          TokBlocking -> do
-            objs <- atomically $ do
-              queue <- readTVar forwardQueue
-              check . not =<< isEmptyTBQueue queue
-              res <- getNTraceObjectsNonBlocking n queue >>= \case
-                []     -> error "impossible"
-                (x:xs) -> return $ x NE.:| xs
-              modifyTVar' wasUsed . const $ True
-              pure res
-            return $ BlockingReply objs
-          TokNonBlocking -> do
-            objs <- atomically $ do
-              queue <- readTVar forwardQueue
-              res <- getNTraceObjectsNonBlocking n queue
-              unless (null res) $
-                modifyTVar' wasUsed . const $ True
-              pure res
-            return $ NonBlockingReply objs
+    { Forwarder.recvMsgTraceObjectsRequest = \blocking (NumberOfTraceObjects n) -> do
+        -- Handle response format outside of `atomically`.
+        res <- atomically $ readFromSinkSTM forwardQueue blocking n
+        pure $ case blocking of
+                 TokNonBlocking -> NonBlockingReply res
+                 TokBlocking    -> BlockingReply $
+                                     -- Convert to a non-empty list.
+                                     case NE.nonEmpty res of
+                                       (Just ne) -> ne
+                                       Nothing -> if n > 0
+                                                  then error "impossible"
+                                                  -- I can't help returning zero
+                                                  -- items in a non-empty list.
+                                                  else error "Good luck!"
```

Comment:
3 small things here:
1. super-optimize the blocking case - it's the most common.
2. `NE.nonEmpty` boxes the value in a lazy `Maybe`. Can you use the same match on the list constructor `(:)` as before - and super-optimize on the non-error case (the old implementation didn't do that)?
3. The error distinction on `n` is somewhat confusing. Can you just use one error msg, that is a little more informative (e.g. "readFromSink: implementation error. STM txn must always block when queue is empty.")